porch/pkg/engine/engine.go

```diff
@@ -417,41 +422,91 @@ func (cad *cadEngine) ListFunctions(ctx context.Context, repositoryObj *configap
 }
 
 type updatePackageMutation struct {
-	task *api.Task
+	oldTask           *api.Task
+	task              *api.Task
+	cad               CaDEngine
+	referenceResolver ReferenceResolver
+	namespace         string
+	pkgName           string
 }
 
 func (m *updatePackageMutation) Apply(ctx context.Context, resources repository.PackageResources) (repository.PackageResources, *api.Task, error) {
 	ctx, span := tracer.Start(ctx, "updatePackageMutation::Apply", trace.WithAttributes())
 	defer span.End()
 
-	// TODO: load directly from source repository
-	dir, err := ioutil.TempDir("", "kpt-pkg-update-*")
+	currUpstreamPkgRef, err := m.currUpstream()
 	if err != nil {
 		return repository.PackageResources{}, nil, err
 	}
-	defer os.RemoveAll(dir)
 
-	if err := writeResourcesToDirectory(dir, resources); err != nil {
-		return repository.PackageResources{}, nil, err
+	targetUpstream := m.task.Clone.Upstream
+	if targetUpstream.Type == api.RepositoryTypeGit || targetUpstream.Type == api.RepositoryTypeOCI {
+		return repository.PackageResources{}, nil, fmt.Errorf("update is not supported for non-porch upstream packages")
 	}
 
-	ref := m.task.Clone.Upstream.Git.Ref
+	originalResources, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchResources(ctx, currUpstreamPkgRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package %s with ref %+v",
+			m.pkgName, *currUpstreamPkgRef)
+	}
 
-	// TODO: This is a hack
-	packageName := filepath.Base(m.task.Clone.Upstream.Git.Directory)
-	packageName = strings.TrimPrefix(packageName, ".git")
+	upstreamRevision, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchRevision(ctx, targetUpstream.UpstreamRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching revision for target upstream %s", targetUpstream.UpstreamRef.Name)
+	}
+	upstreamResources, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchResources(ctx, targetUpstream.UpstreamRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching resources for target upstream %s", targetUpstream.UpstreamRef.Name)
+	}
 
-	packageDir := filepath.Join(dir, packageName)
-	if err := kpt.PkgUpdate(ctx, ref, packageDir, kpt.PkgUpdateOpts{}); err != nil {
-		return repository.PackageResources{}, nil, err
+	klog.Infof("performing pkg upgrade operation for pkg %s resource counts local[%d] original[%d] upstream[%d]",
+		m.pkgName, len(resources.Contents), len(originalResources.Spec.Resources), len(upstreamResources.Spec.Resources))
+
+	// May be have packageUpdater part of engine to make it easy for testing ?
+	updatedResources, err := (&defaultPackageUpdater{}).Update(ctx,
```

Comment:
We need to decide where we draw the line between what should be in the kpt library and what should be done in porch. Given the discussion earlier this week, my thinking is that we should see if we can make the porch layer as thin as possible. I think that means that the update logic will happen based on git or oci, and unaware of the concept of porch packages (which will use either git or oci underneath).

Reply:
I think the challenging part has been mostly around the components that interact with storage (git/oci and credentials) and local filesystem. and we can slowly get there.